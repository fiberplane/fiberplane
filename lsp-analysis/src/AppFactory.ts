import { Hono } from "hono";
import type { RouteElement, RouteTree } from "./types";

export class AppFactory {
  private trees: Array<RouteTree>;
  private apps: Map<string, Hono> = new Map();
  private history: Map<string, RouteElement>;

  constructor(trees: Array<RouteTree>) {
    this.trees = trees;
  }

  public resetHistory() {
    this.history = new Map();
  }

  public hasVisited(id: string): boolean {
    return this.history.has(id);
  }

  public getHistoryLength(): number {
    return this.history.size;
  }

  public getVisitedNode(id: string) {
    return this.trees.find((tree) => tree.id === id);
  }

  public getApp(entryId: string): Hono {
    const root = this.trees.find((tree) => tree.id === entryId);
    this.resetHistory();

    if (!root) {
      throw new Error("Could not find root tree");
    }

    return this.createApp(root);
  }

  private createApp(tree: RouteTree) {
    const app = new Hono();

    this.apps.set(tree.id, app);

    app.use((c, next) => {
      this.history.set(tree.id, tree);
      return next();
    });

    // for (const entry of tree.entries) {
    //   switch (entry.type) {
    //     case "ROUTE_ENTRY": {
    //       app.get(entry.path, (c) => {
    //         this.history.set(entry.id, entry);
    //         return c.text("Ok");
    //       });
    //       continue;
    //     }

    //     case "ROUTE_TREE_REFERENCE": {
    //       if (this.apps.has(entry.targetId)) {
    //         const targetApp = this.apps.get(entry.targetId);
    //         app.route(entry.path, targetApp);
    //       } else {
    //         const tree = this.trees.find((tree) => tree.id === entry.targetId);
    //         if (tree) {
    //           const targetApp = this.createApp(tree);
    //           app.route(entry.path, targetApp);
    //         }
    //       }
    //       continue;
    //     }

    //     case "MIDDLEWARE_ENTRY": {
    //       const middlewareFunc = async (c, next) => {
    //         this.history.set(entry.id, entry);
    //         await next();
    //       };
    //       if (entry.path) {
    //         app.use(entry.path, middlewareFunc);
    //       } else {
    //         app.use(middlewareFunc);
    //       }
    //       continue;
    //     }

    //     default: {
    //       const exhaustiveCheck: never = entry;
    //       // Typescript should never happen
    //       throw new Error(`Unsupported entry type: ${exhaustiveCheck}`);
    //     }
    //   }
    // }

    return app;
  }
}
