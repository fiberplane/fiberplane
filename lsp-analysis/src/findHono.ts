import { inspect, parseArgs } from "node:util";
import type { RouteEntry, RouteTree, TsLanguageService, TsLineAndCharacter, TsNode, TsProgram, TsReferenceEntry, TsSourceFile, TsStringLiteral, TsType } from "./types";
import { isArrowFunction, isImportAttribute } from "typescript";

export function findHonoRoutes(
  server: TsLanguageService,
  ts: TsType,
): {
  errorCount?: number;
  results: Array<RouteTree>;
} {
  const program = server.getProgram();
  if (!program) {
    throw new Error("Program not found");
  }

  let errorCount = 0;
  const checker = program.getTypeChecker();

  const apps: Array<RouteTree> = [];

  // 
  const files = program.getSourceFiles();
  const fileMap: Record<string, TsSourceFile> = {}
  for (const file of files) {
    fileMap[file.fileName] = file;
  }

  for (const sourceFile of files) {
    if (!sourceFile.isDeclarationFile) {
      ts.forEachChild(sourceFile, (file: TsSourceFile) => visit(file, file.fileName));
    }
  }

  function visit(node: TsNode, fileName: string) {
    if (ts.isVariableDeclaration(node) && node.initializer) {
      const type = checker.getTypeAtLocation(node.initializer);
      const typeName = checker.typeToString(type);
      if ("intrinsicName" in type && type.intrinsicName === "error") {
        errorCount++;
        console.error("Error in type check");
        console.debug("In: ", node.getSourceFile().fileName);
        console.debug("Ndoe text:", node.getFullText());

        console.log("type information", type.getSymbol());
      }

      if (typeName.startsWith("Hono<")) {
        const honoInstanceName = node.name.getText();
        // console.log(`Found Hono instance: ${honoInstanceName}`);
        // console.log("parent", node.parent.getText())
        // console.log("parent.parent", node.parent.parent.getText())
        // console.log("parent.parent.parent", node.parent.parent.parent.getText())
        const current: RouteTree = {
          name: honoInstanceName,
          fileName: node.getSourceFile().fileName,
          entries: [],
        };

        // server.getReferencesAtPosition(fileName, arg.getStart());
        findRoutes(current, node.parent.parent.parent);
        apps.push(current);
      }
    }

    ts.forEachChild(node, (child) => visit(child, fileName));
  }

  function findRoutes(routeTree: RouteTree, scope: TsNode) {
    const honoInstanceName = routeTree.name;
    function visit(node: TsNode) {
      if (
        ts.isCallExpression(node) &&
        ts.isPropertyAccessExpression(node.expression)
      ) {
        const methodName = node.expression.name.getText();
        const objectName = node.expression.expression.getText();

        if (
          objectName === honoInstanceName &&
          ["put", "get", "post", "all"].includes(methodName)
        ) {
          const [firstArgument, ...args] = node.arguments;
          const entry: RouteEntry = {
            method: methodName,
            path: JSON.parse(firstArgument.getText()),
            sources: [],
          };

          // Add the tree node to the list of entries
          // Later the entry will be filled with source references
          routeTree.entries.push(entry);

          for (const arg of args) {
            if (ts.isArrowFunction(arg)) {
              // console.log('arrowww')
              const sourceFile = arg.getSourceFile();
              const position = sourceFile.getLineAndCharacterOfPosition(arg.getStart());
              entry.sources.push({
                character: position.character,
                line: position.line,
                fileName: sourceFile.fileName,
                content: arg.getText(),
              });
              // console.log('entry.sources', entry.sources)
            }
            else if (ts.isCallExpression(arg)) {
              const sourceFile = arg.getSourceFile();
              const position = sourceFile.getLineAndCharacterOfPosition(arg.getStart());
              // entry.sources.push({
              //   character: position.character,
              //   line: position.line,
              //   fileName: sourceFile.fileName,
              //   content: arg.getText(),
              // });
              // const sourceFile = arg.getSourceFile();
              const references = server.getReferencesAtPosition(sourceFile.fileName, arg.getStart());
              //   const position = sourceFile.getLineAndCharacterOfPosition(arg.getStart());
              // const isImported = references.find((ref) => {
              //   return ref.fileName.indexOf("/node_modules/") !== -1;
              // });


              // console.log('isImported', isImported, references);
              //   // ts.isImportClause


              //   // let node: TsNode | undefined;
              //   // let
              let attributeNameInfo: undefined | {
                ref: TsReferenceEntry,
                node: TsNode,
              }

              for (const ref of references) {
                if (ref.fileName.indexOf("/node_modules/") === -1) {

                  const currentFile = fileMap[ref.fileName] || program.getSourceFile(ref.fileName);
                  if (!currentFile) {
                    continue;
                  }
                  const position2 = currentFile.getLineAndCharacterOfPosition(ref.textSpan.start);
                  const node = findNodeAtPosition(ts, currentFile, position2);
                  const definitionFileName = getImportTypeDefinitionFileName(ts, node, program);

                  entry.sources.push({
                    character: position.character,
                    line: position.line,
                    fileName: sourceFile.fileName,
                    content: `${definitionFileName}
                    ${arg.getText()}`,
                  });

                  console.log('definitionFileName', definitionFileName, ref)
                  break;
                }
              }

              //   // if (attributeNameInfo) {
              //   //   console.log('importAttributeName', attributeNameInfo.node.parent.getText(), attributeNameInfo.ref);
              //   // }
            }
          }
          // console.log("Arguments", node.arguments.length, node.arguments);
        }
      }

      ts.forEachChild(node, visit);
    }

    ts.forEachChild(scope, visit);
  }

  return {
    results: apps,
    errorCount,
  };
}

function getImportTypeDefinitionFileName(ts: TsType, node: TsNode, program: TsProgram): string | undefined {
  const checker = program.getTypeChecker();
  const symbol = checker.getSymbolAtLocation(node);

  if (symbol) {
    const declarations = symbol.getDeclarations();
    if (declarations && declarations.length > 0) {
      const declaration = declarations[0];
      const sourceFile = declaration.getSourceFile();
      const n = declaration.parent.parent.parent;
      return n.getText();
      // let nextSibling = getNextSibling(n);
      // while (nextSibling && nextSibling.kind !== ts.SyntaxKind.StringLiteral) {
      //   console.log('kind', nextSibling.kind, nextSibling.getText())
      //   nextSibling = getNextSibling(nextSibling);
      // }
      // if (nextSibling) {
      //   console.log('found it', nextSibling.getText())
      //   return nextSibling.getText();
      //   // program.
      //   // const options = ts.
      //   // const something = checker.resolveName(nextSibling.getText(), nextSibling, ts.SymbolFlags., true);
      //   // ts.resolveModuleName(nextSibling.getText(), sourceFile.fileName, {}, {}, undefined, undefined, undefined, undefined);
      //   // console.log('something', something)
      //   // , n, ts.SymbolFlags.Value, ts.SymbolFlags.Value, ts.SymbolFlags.Value, ts.SymbolFlags.Value);
      // }

      // let nextSibling = getNextSibling(n);
      // nextSibling = getNextSibling(nextSibling);
      // checker.resolveName()
      // console.log("-------------------", nextSibling.getFullText())
      // inspectNode(nextSibling, ts)
      // console.log("-------------------")
      // console.log(
      //   sourceFile.getLineAndCharacterOfPosition(declaration.getStart()),
      //   declaration.parent.getFullText())
      // declaration
      // console.log('declaration', declaration.getText(),
      // declarations.length
      // , sourceFile.fileName
      // );
      return sourceFile.fileName;
    }
  }

  return undefined;
}

function getNextSibling(node: TsNode): TsNode | undefined {
  const parent = node.parent;
  if (!parent) {
    // console.log('no parent');
    return undefined;
  }

  const children = parent.getChildren();
  for (let i = 0; i < children.length; i++) {
    if (children[i] === node && i + 1 < children.length) {
      // console.log('found it')
      return children[i + 1];
    }
  }
  // console.log('fallback');

  return undefined;
}
function findNodeAtPosition(ts: TsType, sourceFile: TsSourceFile, lineAndCharacter: TsLineAndCharacter): TsNode | undefined {
  const position = sourceFile.getPositionOfLineAndCharacter(lineAndCharacter.line, lineAndCharacter.character);

  function find(node: TsNode): TsNode | undefined {
    if (position >= node.getStart() && position < node.getEnd()) {
      return ts.forEachChild(node, find) || node;
    }
    return undefined;
  }

  return find(sourceFile);
}

// console.log("node", node.kind, ts.SyntaxKind[node.kind], node.getText());
function inspectNode(node: TsNode, ts: TsType) {
  console.log("node", node.kind, ts.SyntaxKind[node.kind], node.getText());
  console.log("Inspecting node", {
    // isFunctionExpression: ts.isFunctionExpression(node),
    // isArrowFunction: ts.isArrowFunction(node),
    // isFunctionDeclaration: ts.isFunctionDeclaration(node),
    // isFunctionTypeNode: ts.isFunctionTypeNode(node),
    // isCallExpression: ts.isCallExpression(node),
    // isPropertyAccessExpression: ts.isPropertyAccessExpression(node),
    // isIdentifier: ts.isIdentifier(node),
    // isStringLiteral: ts.isStringLiteral(node),
    // isNumericLiteral: ts.isNumericLiteral(node),
    // isObjectLiteralExpression: ts.isObjectLiteralExpression(node),
    // isTypeLiteralNode: ts.isTypeLiteralNode(node),
    // isTypeReferenceNode: ts.isTypeReferenceNode(node),
    // isTypeQueryNode: ts.isTypeQueryNode(node),
    // isTypePredicateNode: ts.isTypePredicateNode(node),
    // // isTypeAssertion: ts.isTypeAssertion(node),
    // isAsExpression: ts.isAsExpression(node),
    // isNonNullExpression: ts.isNonNullExpression(node),
    // isNonNullChain: ts.isNonNullChain(node),
    // isParenthesizedExpression: ts.isParenthesizedExpression(node),
    // isSpreadElement: ts.isSpreadElement(node),
    // isShorthandPropertyAssignment: ts.isShorthandPropertyAssignment(node),
    // isPropertyAssignment: ts.isPropertyAssignment(node),
    // isPropertyDeclaration: ts.isPropertyDeclaration(node),
    // isPropertySignature: ts.isPropertySignature(node),
    // isAccessor: ts.isAccessor(node),
    isImportAttribute: ts.isImportAttribute(node),
    isImportAttributeName: ts.isImportAttributeName(node),
    isImportAttributes: ts.isImportAttributes(node),
    isImportClause: ts.isImportClause(node),
    isImportDeclaration: ts.isImportDeclaration(node),
    isImportEqualsDeclaration: ts.isImportEqualsDeclaration(node),
    isImportOrExportSpecifier: ts.isImportOrExportSpecifier(node),
    isImportSpecifier: ts.isImportSpecifier(node),
    isImportTypeAssertionContainer: ts.isImportTypeAssertionContainer(node),
    isImportTypeNode: ts.isImportTypeNode(node),
  });
}

// function findImportSource(ts: TsType, sourceFile: TsSourceFile, importName: string): string | undefined {
//   let importSource: string | undefined;

//   function visit(node: TsNode) {
//     if (ts.isImportDeclaration(node)) {
//       const importClause = node.importClause;
//       if (importClause && ts.isNamedImports(importClause.namedBindings)) {
//         for (const element of importClause.namedBindings.elements) {
//           if (element.name.getText() === importName) {
//             importSource = (node.moduleSpecifier as TsStringLiteral).text;
//             return;
//           }
//         }
//       }
//     }
//     ts.forEachChild(node, visit);
//   }

//   visit(sourceFile);
//   return importSource;
// }

// Example usage within your existing code
// function findRoutes(routeTree: RouteTree, scope: TsNode) {
//   const honoInstanceName = routeTree.name;

//   function visit(node: TsNode) {
//     if (
//       ts.isCallExpression(node) &&
//       ts.isPropertyAccessExpression(node.expression)
//     ) {
//       const methodName = node.expression.name.getText();
//       const objectName = node.expression.expression.getText();

//       if (
//         objectName === honoInstanceName &&
//         ["put", "get", "post", "all"].includes(methodName)
//       ) {
//         const [firstArgument, ...args] = node.arguments;
//         const entry: RouteEntry = {
//           method: methodName,
//           path: JSON.parse(firstArgument.getText()),
//           sources: [],
//         };

//         // Add the tree node to the list of entries
//         // Later the entry will be filled with source references
//         routeTree.entries.push(entry);

//         for (const arg of args) {
//           if (ts.isArrowFunction(arg)) {
//             const sourceFile = arg.getSourceFile();
//             const position = sourceFile.getLineAndCharacterOfPosition(arg.getStart());
//             entry.sources.push({
//               character: position.character,
//               line: position.line,
//               fileName: sourceFile.fileName,
//               content: arg.getText(),
//             });
//           } else if (ts.isCallExpression(arg)) {
//             console.log("Call expression", arg.getText());
//             const sourceFile = arg.getSourceFile();
//             const references = server.getReferencesAtPosition(sourceFile.fileName, arg.getStart());
//             console.log('references', references)
//             for (const ref of references) {
//               ref
//             }
//           }
//         }
//       }
//     }

//     ts.forEachChild(node, visit);
//   }

//   ts.forEachChild(scope, visit);
// }
