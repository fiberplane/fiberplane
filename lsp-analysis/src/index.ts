// export { getTSServer } from "./server";
export { logger } from "./logger";
export { Watcher } from "./watcher";
export { setupMonitoring } from "./setup";
