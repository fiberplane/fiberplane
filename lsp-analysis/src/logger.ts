import { createConsola } from "consola";

export const logger = createConsola({
  level: 4,
});
