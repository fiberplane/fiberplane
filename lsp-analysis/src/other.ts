import { Hono } from "hono";
export { Hono as App } from "hono";
// export const App = Hono;
const app = new Hono();
console.log("app", app);
// const app2 = new Hono();
// export type SuperHono = Hono & { super: boolean };
// export const App = Hono;
