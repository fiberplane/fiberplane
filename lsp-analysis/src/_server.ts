/**
 * This file contains the TypeScript language server instance and related functions.
 * It uses the `vscode-jsonrpc` library to create a connection to the language server
 * and handles the initialization, notification registration, and cleanup.
 *
 * Note that the language server is initialized in the user's project directory by spawning a child process,
 * and running `npx typescript-language-server --stdio`.
 * We interface with that process via stdin/stdout.
 */

import { type ChildProcessWithoutNullStreams, spawn } from "node:child_process";
import { resolve } from "node:path";
import {
  type MessageConnection,
  StreamMessageReader,
  StreamMessageWriter,
  createMessageConnection,
} from "vscode-jsonrpc/node.js";
import type {
  ConfigurationParams,
  DidChangeConfigurationParams,
  //   // createMessageConnection,
  InitializeParams,
} from "vscode-languageserver-protocol";
import { logger } from "./logger.js";
import { isPublishDiagnosticsParams } from "./types";
import { getFileUri } from "./utils";
// import { connect } from "node:http2";

// let tsServerInstance: {
//   connection: MessageConnection;
//   tsServer: ChildProcessWithoutNullStreams;
// } | null = null;

/**
 * Retrieves the TypeScript language server instance for a given project path.
 * If an instance already exists, it will be reused to avoid unnecessary initialization.
 *
 * @param {string} pathToProject - The path to the project directory.
 * @returns {Promise<{ connection: MessageConnection; tsServer: ChildProcessWithoutNullStreams }>} The TypeScript language server instance.
 */
export async function getTSServer(pathToProject: string) {
  const resolvedPath = resolve(pathToProject);

  // if (tsServerInstance) {
  //   // logger.debug(
  //   //   chalk.dim(
  //   //     `[debug] Reusing existing TS Server instance for project: ${resolvedPath}`,
  //   //   ),
  //   // );
  //   return tsServerInstance;
  // }

  return await initializeTSServer(resolvedPath);
  // return tsServerInstance;
}

async function initializeTSServer(pathToProject: string) {
  logger.debug(`Initializing TS Server for project: ${pathToProject}`);

  const tsServer = spawn(
    "npx",
    [
      "typescript-language-server",
      "--log-level",
      "verbose",
      // "--logVerbosity=verbose",
      "--stdio",
      // "--tsserver-log-file",
      // "out.log",
    ],
    {
      // NOTE - This will add quite a bit of startup time if the user has not yet downloaded typescript-language-server dependency before via npx...
      //        And I haven't tested what that overhead is...
      // NOTE - We *could* run the language server in the project root (via cwd switch), but I think we handle this by initializing the language server
      //        and pointing it to the project root below anyhow.
      // cwd: pathToProject,
      shell: true,
    },
  );

  // Terminate the language server when the Node.js process exits
  process.on("exit", () => {
    logger.debug(
      `[debug] Terminating TS Server instance for project: ${pathToProject}`,
    );
    tsServer.kill();
  });

  // NOTE - Uncomment to debug raw output of ts-language-server
  //
  tsServer.stdout.on("data", (data) => {
    // console.log("<TS-SERVER-DATA>");
    console.log(`--> tsServer stdout: ${data.toString()}`);
    // logger.box(`tsServer stdout: ${data.toString()}`);
    // console.log("</TS-SERVER-DATA>");
  });

  // tsServer.stderr.on("data", (data) => {
  //   logger.error(`tsserver stderr: ${data.toString()}`);
  // });

  tsServer.stdout.on("data", (data) => {
    logger.trace(`tsserver stdout: ${data.toString()}`);
  });

  const connection = createMessageConnection(
    new StreamMessageReader(tsServer.stdout),
    new StreamMessageWriter(tsServer.stdin),
  );

  connection.listen();

  tsServer.on("close", (code) => {
    logger.debug(`typescript-language-server process exited with code ${code}`);
  });

  try {
    const rootUri = getFileUri(pathToProject);
    logger.debug(
      `Initializing typescript language server with rootUri: ${rootUri}`,
    );

    // ClientCapabilities
    const params: InitializeParams = {
      processId: process.pid,
      rootUri: rootUri,
      capabilities: {
        workspace: {
          // codeLens: {
          //   // dynamicRegistration: true,
          //   refreshSupport: true,
          // },
          configuration: true,
        },
      },
      workspaceFolders: [{ uri: rootUri, name: "app" }],
      trace: "verbose",

      initializationOptions: {
        typescript: {
          referencesCodeLens: {
            enabled: true,
          },
          implementationsCodeLens: {
            enabled: true,
          },
        },

        preferences: {
          allowIncompleteCompletions: true,
          includeCompletionsForModuleExports: true,
          includeCompletionsWithInsertText: true,
        },
      },
    };

    const ir = await connection.sendRequest("initialize", params);
    console.log("ir", ir);
    await connection.sendNotification("initialized");

    registerNotificationHandlers(connection);
    await new Promise((resolve) => setTimeout(resolve, 200));
    // enable codelens
    const configParams: DidChangeConfigurationParams = {
      settings: {
        asdasdfasd: "erwe",
        // typescript: {
        referencesCodeLens: {
          enabled: true,
        },
        implementationsCodeLens: {
          enabled: true,
        },
        // },
      },
    };
    // const dcc =
    await connection.sendNotification(
      "workspace/didChangeConfiguration",
      configParams,
    );
    // console.log("configUpdate dcc ", dcc);

    // const configQueryParams: ConfigurationParams = {
    //   items: [
    //     {
    //       section: "typescript.referencesCodeLens",
    //     },
    //     {
    //       section: "typescript.implementationsCodeLens",
    //     },
    //   ],
    // };
    // const r = await connection.sendRequest(
    //   "workspace/configuration",
    //   configQueryParams,
    // );
    // console.log("configUpdate", r);

    return { connection, tsServer };
  } catch (error) {
    logger.error("Error initializing TS Server:", error);
    throw error;
  }
}

/**
 * Registers handlers for various notifications from the TypeScript language server.
 *
 * @param {MessageConnection} connection - The connection to the TypeScript language server.
 */
function registerNotificationHandlers(connection: MessageConnection) {
  const logger = console;
  connection.onNotification("textDocument/didOpen", (params) => {
    logger.debug(
      `[textDocument/didOpen] Received definition request: ${JSON.stringify(
        params,
        null,
        2,
      )}`,
    );
  });
  // Listen for diagnostics
  connection.onNotification(
    "textDocument/publishDiagnostics",
    (params: unknown) => {
      if (!isPublishDiagnosticsParams(params)) {
        logger.debug(
          "[debug] Received unexpected params for `textDocument/publishDiagnostics`",
        );
        return;
      }
      const { uri, diagnostics } = params;
      if (diagnostics.length > 0) {
        logger.info(
          `[textDocument/publishDiagnostics] Diagnostics for ${uri}:`,
        );
        for (const diag of diagnostics) {
          logger.info(
            `- [${diag.severity}] ${diag.message} at ${diag.range.start.line}:${diag.range.start.character}`,
          );
        }
      } else {
        logger.debug(
          `[textDocument/publishDiagnostics] No diagnostics for ${uri}.`,
        );
      }
    },
  );

  // Listen for log messages
  connection.onNotification("window/logMessage", (params) => {
    const { type, message } = params;
    logger.info(`[window/logMessage] Log Message [${type}]: ${message}`);
    logger.debug(params);
  });

  // Listen for show messages
  connection.onNotification("window/showMessage", (params) => {
    const { type, message } = params;
    logger.debug(`[window/showMessage] Show Message [${type}]: ${message}`);
  });
}

// type
