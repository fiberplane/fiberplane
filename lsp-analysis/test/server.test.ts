import {
  // assert,
  expect,
  test,
  // vi
} from "vitest";
// import * as fs from "node:fs";
// import { findHonoLikeSymbols, setup } from "../src";
// import { Watcher } from "../src/monitor";
import * as path from "node:path";
import { setupMonitoring } from "src";
// import { getTSServer, logger } from "../src";
// import type { MessageConnection } from "vscode-jsonrpc";
// import type {
//   // CodeLensParams,
//   Definition,
//   DidOpenTextDocumentParams,
//   Location,
//   TextDocumentItem,
// } from "vscode-languageserver-protocol";
// import {
//   codeLens,
//   getTsSourceDefinition,
//   getWordAtLocation,
//   // sendDefinitionRequest,
//   sendReferencesRequest,
// } from "./utils";

test("single file", async () => {
  const location = path.join(__dirname, "./test-case/single");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(location);
  try {
    watcher.start();

    expect(findHonoRoutes()).toMatchSnapshot()

  } finally {
    teardown();
  }
});

test("multiple files", async () => {
  const location = path.join(__dirname, "./test-case/multiple");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(location);
  try {
    watcher.start();

    expect(findHonoRoutes()).toMatchSnapshot()
  } finally {
    teardown();
  }
});

test("module imports", async () => {
  const location = path.join(__dirname, "./test-case/module-imports");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(location);
  try {
    watcher.start();

    expect(findHonoRoutes()).toMatchSnapshot()
  } finally {
    teardown();
  }
});

// test("goose quotes", async () => {
//   const location = path.resolve(path.join(__dirname, "../../api"));
//   const { watcher, findHonoRoutes, teardown } = setupMonitoring(location);
//   try {
//     watcher.start();

//     const result = findHonoRoutes()
//     console.log('result', JSON.stringify(result, null, 2))
//     expect(result).toMatchSnapshot()
//   } finally {
//     teardown();
//   }

// });

//   const location = path.join(__dirname, "./test-case/multiple");
//   const { connection, tsServer } = await getTSServer(location);

//   const otherFilePath = path.resolve(location, "other.ts");
//   await didOpen(connection, otherFilePath, location);
//   const indexFilePath = path.resolve(location, "index.ts");
//   await didOpen(connection, indexFilePath, location);
//   console.log("wait before codelens")
//   await new Promise((resolve) => setTimeout(resolve, 1000));

//   await codeLens(connection, indexFilePath);
//   await new Promise((resolve) => setTimeout(resolve, 100));
//   const definition = await checkFile(connection, indexFilePath, location);
//   if (Array.isArray(definition)) {
//     printResults(definition);
//   } else if (definition) {
//     printElement(definition);
//   }
//   tsServer.kill();
// });

// function printElement(element: Location) {
//   const content = data[element.uri]?.text;
//   console.log("element", element.uri, !!content);
//   if (!content) {
//     return;
//   }

//   const word = getWordAtLocation(content, element.range.start);
//   console.log("element", element.uri, word, element.range.start);
// }

// function printResults(locations: Array<Location>) {
//   // console.log("location", locations.length);
//   for (const result of locations) {
//     printElement(result);
//   }
// }

// async function checkFile(
//   connection: MessageConnection,
//   filePath: string,
//   root: string,
// ) {
//   // const fileUri = `file://${filePath.replace(root, "")}`;
//   const fileUri = `file://${filePath}`;
//   const position = {
//     line: 0,
//     character: 9,
//   };

//   const document = data[fileUri];
//   const content = document.text;
//   const word = getWordAtLocation(content, position);

//   console.debug(`looking for \`${word}\` in ${fileUri}`);
//   // return await sendReferencesRequest(connection, fileUri, position);
//   // return await sendReferencesRequest(connection, fileUri, position);
//   return getTsSourceDefinition(connection, fileUri, position);
// }

// const data: Record<string, TextDocumentItem> = {};

// function didOpen(
//   connection: MessageConnection,
//   location: string,
//   root: string,
// ) {
//   const content = fs.readFileSync(location, "utf-8");
//   // const fileUri = `file://${location.replace(root, "")}`;
//   const fileUri = `file://${location}`;
//   data[fileUri] = {
//     uri: fileUri,
//     languageId: "typescript",
//     version: 1,
//     text: content,
//   };

//   const params: DidOpenTextDocumentParams = {
//     textDocument: data[fileUri],
//   };

//   return connection.sendNotification("textDocument/didOpen", params);
// }

// // function makeCodeLensRequest(connection: MessageConnection, fileUri: string) {
// //   const params: CodeLensParams = {
// //     textDocument: {
// //       uri: fileUri,
// //     },

// //   }
// //   return connection.sendRequest("textDocument/codeLens", params);
// // }
