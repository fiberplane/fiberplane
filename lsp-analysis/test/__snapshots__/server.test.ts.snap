// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`module imports 1`] = `
{
  "errorCount": 0,
  "results": [
    {
      "entries": [
        {
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 2,
              "content": "import { cors } from "hono/cors";
                    cors()",
              "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/module-imports/index.ts",
              "line": 7,
            },
            {
              "character": 2,
              "content": "() => cors()",
              "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/module-imports/index.ts",
              "line": 8,
            },
            {
              "character": 2,
              "content": "(c) => c.text("Hello, Hono!")",
              "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/module-imports/index.ts",
              "line": 9,
            },
          ],
        },
      ],
      "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/module-imports/index.ts",
      "name": "app",
    },
  ],
}
`;

exports[`multiple files 1`] = `
{
  "errorCount": 0,
  "results": [
    {
      "entries": [
        {
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 13,
              "content": "(c) => c.text("Hello, Hono!")",
              "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/multiple/index.ts",
              "line": 4,
            },
          ],
        },
      ],
      "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/multiple/index.ts",
      "name": "app",
    },
    {
      "entries": [
        {
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 13,
              "content": "(c) => {
  return c.json({});
}",
              "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/multiple/other.ts",
              "line": 2,
            },
          ],
        },
        {
          "method": "post",
          "path": "/",
          "sources": [
            {
              "character": 14,
              "content": "(c) => c.json({ "hello": "world" })",
              "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/multiple/other.ts",
              "line": 5,
            },
          ],
        },
      ],
      "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/multiple/other.ts",
      "name": "app",
    },
  ],
}
`;

exports[`single file 1`] = `
{
  "errorCount": 0,
  "results": [
    {
      "entries": [
        {
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 13,
              "content": "(c) => c.text("Hello, Hono!")",
              "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/single/index.ts",
              "line": 4,
            },
          ],
        },
      ],
      "fileName": "/Users/jaccoflenter/dev/fiberplane/fpx/lsp-analysis/test/test-case/single/index.ts",
      "name": "app",
    },
  ],
}
`;
